File: android/app/src/main/java/com/connect/club/react/bridge/iosLike/ui/go/HttpClient.kt
package com.connect.club.react.bridge.iosLike.ui.go

import android.os.Build
import com.connect.club.BuildConfig
import com.connect.club.react.bridge.iosLike.debugP
import com.connect.club.react.bridge.iosLike.LogUtil
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import common.Common
import common.PublicHttpClient
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

private const val AMPLITUDE_SESSION_ID = "amplitudeSessionId"
private const val AMPLITUDE_DEVICE_ID = "amplitudeDeviceId"

class HttpClient(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName() = "HttpClient"

    private val goAsyncStorage: GoAsyncStorage
    private var httpClient: PublicHttpClient? = null

    init {
        goAsyncStorage = GoAsyncStorage(reactContext)
        goAsyncStorage.delete(AMPLITUDE_SESSION_ID)
        goAsyncStorage.delete(AMPLITUDE_DEVICE_ID)
        Common.setStorage(goAsyncStorage)
    }

    @ReactMethod
    fun initialize(endpoint: String, promise: Promise) {
        debugP("HttpClient.initialize")
        httpClient = Common.httpClient(
                endpoint,
                "android",
                BuildConfig.VERSION_NAME,
                BuildConfig.VERSION_CODE.toString(),
                Build.VERSION.SDK_INT.toString()
        )
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.N) {
            Common.insecureHttpTransport()
        }
        promise.resolve(true)
    }

    @ReactMethod
    fun setAmplitudeIds(sessionId: String, deviceId: String) {
        goAsyncStorage.setString(AMPLITUDE_SESSION_ID, sessionId)
        goAsyncStorage.setString(AMPLITUDE_DEVICE_ID, deviceId)
    }

    @ReactMethod
    fun isAuthorized(promise: Promise) {
        promise.resolve(goAsyncStorage.getString("accessToken").isNotEmpty())
    }

    @ReactMethod
    fun queryAuthorize(
        query: String,
        promise: Promise
    ) {
        GlobalScope.launch {
            val response = httpClient!!.authorize(query)
            promise.resolve(response)
        }
    }

    @ReactMethod
    fun request(
            endpoint: String,
            method: String,
            useAuthorizeHeader: Boolean,
            generateJwt: Boolean,
            query: String?,
            body: String?,
            file: String?,
            promise: Promise
    ) {
        GlobalScope.launch {
            val response = httpClient!!.getRequest(endpoint, method, useAuthorizeHeader, generateJwt, query, body, "photo", "photo", file)
            promise.resolve(response)
        }
    }

    @ReactMethod
    fun sendLogFile(body: String, promise: Promise) {
        GlobalScope.launch {
            val logInfo = "${LogUtil.getDeviceInfoBody()}\n${body}"
            val response = httpClient!!.sendLogFileWithBodyText(logInfo)
            promise.resolve(response)
        }
    }

    @ReactMethod
    fun sendPreservedLogFile(body: String, promise: Promise) {
        GlobalScope.launch {
            val logInfo = "(Error) ${LogUtil.getDeviceInfoBody()}\n${body}"
            val response = httpClient!!.sendLogFileWithPath(LogUtil.preservedLogFilePath, logInfo)
            promise.resolve(response)
        }
    }
}

File: ios/ReactNativeModules/HttpClient.swift
//
//  HttpClient.swift
//  connectreactive
//
//  Created by Sergei Golishnikov on 18/05/2021.
//

import Foundation
import React
import Common
import Kingfisher

@objc(HttpClient)
final class HttpClient: RCTEventEmitter {

    static var logInfo = ""
    static var httpClient: CommonPublicHttpClientProtocol?

    private var AMPLITUDE_SESSION_ID = "amplitudeSessionId"
    private var AMPLITUDE_DEVICE_ID = "amplitudeDeviceId"

    let queue = DispatchQueue(label: "httpQueue")
    let goAsyncStorage = GoAsyncStorage.shared

    override init() {
        super.init()
        CommonSetStorage(goAsyncStorage)
        goAsyncStorage.delete(AMPLITUDE_SESSION_ID)
        goAsyncStorage.delete(AMPLITUDE_DEVICE_ID)
    }

    override func supportedEvents() -> [String]! { [] }
    override class func requiresMainQueueSetup() -> Bool {
        false
    }

    @objc
    func initialize(
        _ endpoint: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as! String
        let versionName = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as! String
        let buildNumber = UIDevice.current.systemVersion

        Self.logInfo = String(
            format: "iOS %@ build: %@(%@) device: %@",
            UIDevice.current.systemVersion,
            version,
            versionName,
            UIDevice.deviceType.rawValue
        )

        Self.httpClient = CommonHttpClient(
            endpoint,
            "ios",
            version,
            versionName,
            buildNumber
        )
        resolver(true)

        #if DEBUG
        return
        #endif
    }

    @objc
    func isAuthorized(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock) {
        resolver(!goAsyncStorage.getString("accessToken").isEmpty)
    }

    @objc
    func queryAuthorize(
        _ query: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        queue.async {
            let response = Self.httpClient!.authorize(query)
            resolver(response)
        }
    }

    @objc
    func request(
        _ endpoint: String,
        method: String,
        useAuthorizeHeader: Bool,
        generateJwt: Bool,
        query: String?,
        body: String?,
        file: String?,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        queue.async {
            let response = Self.httpClient!.getRequest(
                endpoint,
                method: method,
                useAuthorizeHeader: useAuthorizeHeader,
                generateJwt: generateJwt,
                query: query,
                body: body,
                filePartName: "photo",
                fileName: "photo",
                filePath: file
            )
            resolver(response)
        }
    }

    @objc
    func sendLogFile(
        _ body: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        let text = "\(Self.logInfo)\n\(body)"
        let response = Self.httpClient!.sendLogFile(withBodyText: text)
        resolver(response)
    }

    @objc
    func sendPreservedLogFile(
        _ body: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        let path = Logger.preservedLogFilePath
        let text = "(Error) \(Self.logInfo)\n\(body)"
        let response = Self.httpClient!.sendLogFile(withPath: path, bodyText: text)
        resolver(response)
    }

    @objc
    func setAmplitudeIds(_ sessionId: String, deviceId: String) {
        goAsyncStorage.setString(AMPLITUDE_SESSION_ID, value: sessionId)
        goAsyncStorage.setString(AMPLITUDE_DEVICE_ID, value: deviceId)
    }
}
